### PR TITLE
[Win] Add functions to access bundle resources

### PR DESCRIPTION
**Pull request by donny-dont**
#### e5946f5567d06386d9d43168a6f688144840aab5
<pre>
[Win] Add functions to access bundle resources
<a href="https://bugs.webkit.org/show_bug.cgi?id=244096">https://bugs.webkit.org/show_bug.cgi?id=244096</a>

Reviewed by Darin Adler.

The Windows ports have a `WebKit.resources` directory which contains assets, such as the
WebInspector UI, that are used by the engine. WebCoreBundle has a function to get a `CFBundle` that
is then queried to get the actual resource. Instead of using the bundle directly functions are
added that request the location of the resource. When `USE(CF)` these functions use the `CFBundle`
to determine the location, otherwise `WTF::FileSystem` is used to get the location.

This is another patch that gets the WinCairo build closer to removing CF usage.

* Source/WebCore/platform/win/WebCoreBundleWin.cpp:
(WebCore::webKitBundlePath):
(WebCore::dllDirectory):
* Source/WebCore/platform/win/WebCoreBundleWin.h:
* Source/WebCore/rendering/RenderThemeWin.cpp:
(WebCore::RenderThemeWin::stringWithContentsOfFile):
* Source/WebKit/UIProcess/Inspector/win/InspectorResourceURLSchemeHandler.cpp:
(WebKit::InspectorResourceURLSchemeHandler::platformStartTask):
* Source/WebKit/UIProcess/win/WebProcessPoolWin.cpp:
(WebKit::initializeRemoteInspectorServer):
* Tools/TestWebKitAPI/PlatformWin.cmake:
* Tools/TestWebKitAPI/Tests/WebCore/win/WebCoreBundle.cpp: Added.

Canonical link: <a href="https://commits.webkit.org/253925@main">https://commits.webkit.org/253925@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/32c5d182e18ed505d9153693ad685cea440fc423

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87465 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31554 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18270 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/96692 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/150102 "Found 1 new test failure: fast/css/aspect-ratio-min-height-replaced.html") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/91440 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/29919 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/26101 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79583 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91463 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93083 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/24174 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/74250 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/24005 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/79141 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79362 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/67023 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27629 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/13194 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27582 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14210 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2756 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/29271 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/37073 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/29198 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33488 "Passed tests") | |

### CODE REVIEW

File: Source/WTF/wtf/URL.cpp [hunk 1]
```diff
@@ -250,11 +250,7 @@ String URL::fileSystemPath() const
     if (!isLocalFile())
         return { };
 
-    auto result = decodeEscapeSequencesFromParsedURL(path());
-#if PLATFORM(WIN)
-    result = FileSystem::fileSystemRepresentation(result);
-#endif
-    return result;
+    return decodeEscapeSequencesFromParsedURL(path());
```
Comment (foopoiuyt) [inline Source/WTF/wtf/URL.cpp]: Since this is a general method - does this change work for all the users of this method (some of which might not necessarily be obvious just from a simple does a page load type test)?

File: Source/WebKit/UIProcess/Inspector/win/InspectorResourceLocation.cpp [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+#if ENABLE(REMOTE_INSPECTOR)
+
+#include <wtf/FileSystem.h>
+
+namespace WebKit {
+
+static String bundlePath()
+{
+    wchar_t buffer[MAX_PATH];
+    DWORD length = ::GetModuleFileNameW(nullptr, buffer, WTF_ARRAY_LENGTH(buffer));
```
Comment (darinadler) [inline Source/WebKit/UIProcess/Inspector/win/InspectorResourceLocation.cpp]: New code should use `std::size` instead of `WTF_ARRAY_LENGTH`.

File: Source/WebKit/UIProcess/Inspector/win/InspectorResourceLocation.cpp [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+namespace InspectorResource {
+
+String backendCommandsPath()
+{
+    return FileSystem::pathByAppendingComponents(bundlePath(), { "WebKit.Resources"_s, "WebInspectorUI"_s, "Protocol"_s, "InspectorBackendCommands.js"_s });
+}
+
+String uiPath(StringView resource)
```
Comment (darinadler) [inline Source/WebKit/UIProcess/Inspector/win/InspectorResourceLocation.cpp]: I assume the code is tested, so these questions must already have sensible answers, but I am wondering: Is the `resource` argument allowed to have separators in it? If so, are they `/` characters or `\` characters on Windows? Does the `pathByAppendingComponents` implementation support arguments with separators in them?

File: Source/WebKit/UIProcess/Inspector/win/InspectorResourceLocation.h [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#if ENABLE(REMOTE_INSPECTOR)
+
+#include <wtf/text/WTFString.h>
```
Comment (darinadler) [inline Source/WebKit/UIProcess/Inspector/win/InspectorResourceLocation.h]: Should only need `Forward.h` for the declarations below, not `WTFString.h`.

File: Source/WebKit/UIProcess/Inspector/win/InspectorResourceURLSchemeHandler.cpp [hunk 5]
```diff
@@ -35,24 +36,25 @@
...
+    if (requestPath.startsWith("/"_s))
         requestPath = requestPath.substring(1);
-    auto path = URL(adoptCF(CFBundleCopyBundleURL(WebCore::webKitBundle())).get()).fileSystemPath();
-    path = FileSystem::pathByAppendingComponent(path, "WebInspectorUI"_s);
-    path = FileSystem::pathByAppendingComponent(path, requestPath);
+    auto path = InspectorResource::uiPath(requestPath);
     bool success;
     FileSystem::MappedFileData file(path, FileSystem::MappedFileMode::Private, success);
     if (!success) {
+#if USE(CURL)
```
Comment (darinadler) [inline Source/WebKit/UIProcess/Inspector/win/InspectorResourceURLSchemeHandler.cpp]: Why this conditional? The file already does `#error` above `if !USE(CURL)`. Is this file intended to compile in non-`USE(CURL)` cases, like in the Apple Windows platform?

File: Source/WTF/wtf/URL.cpp (hunk 1 above)
Comment (darinadler) [reply Source/WTF/wtf/URL.cpp]: Further: Why do we need to remove this?

**Review by darinadler — COMMENTED** [review]
I’m not sure whether to approve this or not. Can’t figure out if it’s right yet.

**Comment by Constellation** [comment]
NICE!

File: Source/WebKit/UIProcess/Inspector/win/InspectorResourceLocation.h (hunk 4 above)
Comment (donny-dont) [reply Source/WebKit/UIProcess/Inspector/win/InspectorResourceLocation.h]: File removed so resolving

File: Source/WebKit/UIProcess/Inspector/win/InspectorResourceLocation.cpp (hunk 3 above)
Comment (donny-dont) [reply Source/WebKit/UIProcess/Inspector/win/InspectorResourceLocation.cpp]: File removed so resolving

File: Source/WTF/wtf/URL.cpp (hunk 1 above)
Comment (donny-dont) [reply Source/WTF/wtf/URL.cpp]: I'm walking back this change for the moment and will deal with it in a separate commit

File: Source/WebKit/UIProcess/Inspector/win/InspectorResourceURLSchemeHandler.cpp (hunk 5 above)
Comment (donny-dont) [reply Source/WebKit/UIProcess/Inspector/win/InspectorResourceURLSchemeHandler.cpp]: A hypothetical Apple Win COULD use this file if so desired. The part thats curl specific is the error being reported.

Comment (darinadler) [reply Source/WebKit/UIProcess/Inspector/win/InspectorResourceURLSchemeHandler.cpp]: We don’t need #if statements for hypothetical ports. I suggest leaving the conditional out.

File: Source/WebCore/platform/win/WebCoreBundleWin.h [hunk 6]
```diff
@@ -23,19 +24,23 @@
...
-
-}
-
 #endif
 
-#endif // WebCoreBundleWin_h
+String webKitBundlePath();
+String webKitBundlePath(StringView);
+String webKitBundlePath(StringView, StringView, StringView);
+String webKitBundlePath(const Vector<StringView>&);
```
Comment (darinadler) [inline Source/WebCore/platform/win/WebCoreBundleWin.h]: Ideally should take a `Span<const StringView>` instead. That‘s something you can use to pass a `Vector` of `StringView` or an array of `StringView` and we should be using that for future functions that take a vector that it looks at. But I suppose we’d also have to change `FileSystem::pathByAppendingComponents` if we wanted to fix this.

File: Source/WebCore/platform/win/WebCoreBundleWin.h [hunk 7]
```diff
@@ -23,19 +24,23 @@
...
 CFBundleRef webKitBundle();
-
-}
-
 #endif
 
-#endif // WebCoreBundleWin_h
+String webKitBundlePath();
+String webKitBundlePath(StringView);
+String webKitBundlePath(StringView, StringView, StringView);
```
Comment (darinadler) [inline Source/WebCore/platform/win/WebCoreBundleWin.h]: This function isn’t much like the others. It has arguments with special meaning. At the very least we need argument names here. But maybe this function also should have its own distinct name.

File: Source/WebCore/platform/win/WebCoreBundleWin.cpp [hunk 8]
```diff
@@ -61,6 +66,92 @@ CFBundleRef webKitBundle()
...
+}
 
-#endif
+String webKitBundlePath(StringView path)
+{
+    auto pathString = path.toStringWithoutCopying();
+    auto directory = FileSystem::parentPath(pathString);
+    auto fileName = FileSystem::pathFileName(pathString);
+
+    auto parts = fileName.split('.');
```
Comment (darinadler) [inline Source/WebCore/platform/win/WebCoreBundleWin.cpp]: We don’t need to do a split, since we are really just trying to shave off the last period-separated piece. We should use `reverseFind` and then we can use `substring`, `left`, and `right`.

If we do want to stick with `split`, then it could be more efficient if we use the `split` from `StringView` rather than the one from `String`.

File: Source/WebCore/platform/win/WebCoreBundleWin.cpp [hunk 9]
```diff
@@ -61,6 +66,92 @@ CFBundleRef webKitBundle()
...
+}
+
+#else
+
+static String dllDirectory()
+{
+    WCHAR buffer[MAX_PATH];
+    DWORD length = ::GetModuleFileNameW(WebCore::instanceHandle(), buffer, MAX_PATH);
+    if (!length || (length == MAX_PATH && GetLastError() == ERROR_INSUFFICIENT_BUFFER))
+        return { };
```
Comment (darinadler) [inline Source/WebCore/platform/win/WebCoreBundleWin.cpp]: It is really OK to just use an empty string here? Maybe there is some other best practice needed here.

File: Source/WebCore/platform/win/WebCoreBundleWin.cpp [hunk 10]
```diff
@@ -61,6 +66,92 @@ CFBundleRef webKitBundle()
...
+    if (name == "mediaControlsLocalizedStrings"_s)
+        return webKitBundlePath(FileSystem::pathByAppendingComponent("en.lproj"_s, fileName));
+
+    return webKitBundlePath(FileSystem::pathByAppendingComponent(directory, fileName));
+}
+
+String webKitBundlePath(const Vector<StringView>& components)
+{
+    return webKitBundlePath(FileSystem::pathByAppendingComponents(""_s, components));
+}
```
Comment (darinadler) [inline Source/WebCore/platform/win/WebCoreBundleWin.cpp]: Seems like this should be outside the conditionals. It’s a function that should work as-is on all platforms.

File: Source/WebCore/platform/win/WebCoreBundleWin.cpp [hunk 11]
```diff
@@ -61,6 +66,92 @@ CFBundleRef webKitBundle()
...
+    auto resource = FileSystem::pathByAppendingComponent(webKitBundlePath(), path);
+    if (!FileSystem::fileExists(resource))
+        return emptyString();
+
+    return resource;
+}
+
+String webKitBundlePath(StringView name, StringView type, StringView directory)
+{
+    auto fileName = makeString(name, "."_s, type);
```
Comment (darinadler) [inline Source/WebCore/platform/win/WebCoreBundleWin.cpp]: More efficient to use `'.'` instead of `"."_s`.

File: Source/WebCore/platform/win/WebCoreBundleWin.cpp [hunk 12]
```diff
@@ -61,6 +66,92 @@ CFBundleRef webKitBundle()
...
+{
+    static NeverDestroyed<String> bundle = FileSystem::pathByAppendingComponent(dllDirectory(), "WebKit.resources"_s);
+    return bundle;
+}
+
+String webKitBundlePath(StringView path)
+{
+    auto resource = FileSystem::pathByAppendingComponent(webKitBundlePath(), path);
+    if (!FileSystem::fileExists(resource))
+        return emptyString();
```
Comment (darinadler) [inline Source/WebCore/platform/win/WebCoreBundleWin.cpp]: Do callers know how to deal with this empty string? How did we choose empty string rather than null string?

File: Source/WebCore/platform/win/WebCoreBundleWin.h [hunk 13]
```diff
@@ -23,19 +24,23 @@
...
+#if USE(CF)
 CFBundleRef webKitBundle();
-
-}
-
 #endif
 
-#endif // WebCoreBundleWin_h
+String webKitBundlePath();
+String webKitBundlePath(StringView);
```
Comment (darinadler) [inline Source/WebCore/platform/win/WebCoreBundleWin.h]: We need an argument name here. It’s not obvious what this argument is.

File: Source/WebCore/platform/win/WebCoreBundleWin.cpp [hunk 14]
```diff
@@ -61,6 +66,92 @@ CFBundleRef webKitBundle()
...
+    auto resourceURL = adoptCF(CFBundleCopyResourceURL(webKitBundle(), name.createCFString().get(), type.createCFString().get(), directory.createCFString().get()));
+    if (!resourceURL)
+        return { };
+
+    return adoptCF(CFURLCopyFileSystemPath(resourceURL.get(), kCFURLWindowsPathStyle)).get();
+}
+
+String webKitBundlePath(const Vector<StringView>& components)
+{
+    return webKitBundlePath(FileSystem::pathByAppendingComponents(""_s, components));
```
Comment (darinadler) [inline Source/WebCore/platform/win/WebCoreBundleWin.cpp]: We have a function `emptyString()` that I believe is always more efficient than `""_s`. Or maybe we should change `""_s` to generate the exact same code as `emptyString()`?

Applies to all the places we use `""_s`.

**Review by darinadler — APPROVED** [review]

**Comment by brentfulgham** [comment]
I think this patch looks good, fwiw.
